pkg/kapis/events/v1alpha2/handler.go

```diff
@@ -0,0 +1,191 @@
+package v1alpha2
+
+import (
+	"errors"
+	"fmt"
+	"github.com/emicklei/go-restful"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/klog"
+	"kubesphere.io/kubesphere/pkg/api"
+	apiEvents "kubesphere.io/kubesphere/pkg/api/events/v1alpha2"
+	"kubesphere.io/kubesphere/pkg/apis/tenant/v1alpha1"
+	"kubesphere.io/kubesphere/pkg/apiserver/query"
+	"kubesphere.io/kubesphere/pkg/apiserver/request"
+	"kubesphere.io/kubesphere/pkg/constants"
+	"kubesphere.io/kubesphere/pkg/informers"
+	"kubesphere.io/kubesphere/pkg/models/events/v1alpha2"
+	"kubesphere.io/kubesphere/pkg/models/tenant"
+	"kubesphere.io/kubesphere/pkg/simple/client/events"
+	"kubesphere.io/kubesphere/pkg/utils/stringutils"
+	"strconv"
+	"strings"
+	"time"
+)
+
+type handler struct {
+	tenant tenant.Interface
+	eo     v1alpha2.EventsOperator
+}
+
+func newHandler(factory informers.InformerFactory, ec events.Client) *handler {
+	return &handler{tenant: tenant.New(factory), eo: v1alpha2.NewEventsOperator(ec)}
+}
+
+func (h handler) Query(req *restful.Request, resp *restful.Response) {
+	suser, ok := request.UserFrom(req.Request.Context())
+	if !ok {
+		err := errors.New("cannot obtain user info")
+		klog.Errorln(err)
+		api.HandleForbidden(resp, req, err)
+		return
+	}
+	userWorkspaces, userNamespaces, err := h.userScope(suser)
+	if err != nil {
+		klog.Errorln(err)
+		api.HandleInternalError(resp, req, err)
+		return
+	}
+	filter, err := h.parseFilterFrom(req, userWorkspaces, userNamespaces)
+	if err != nil {
+		klog.Errorln(err)
+		api.HandleInternalError(resp, req, err)
+		return
+	}
+
+	var ar apiEvents.APIResponse
+	if len(filter.InvolvedObjectNamespaceMap) == 0 { // does not query and returns empty result
```

Comment:
this code snippet is to limit access of events to specific ns?  as I can remember there are events that have no namespace? For these events, admin should be able to search